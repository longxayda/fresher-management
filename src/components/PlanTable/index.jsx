import React from "react";
import {
  useTable,
  usePagination,
  useGlobalFilter,
  useSortBy,
  useFilters,
  useRowSelect,
} from "react-table";
import { Input, Label } from "reactstrap";
import { Button, Table, Form ,Spinner} from "react-bootstrap";

function NoneColumnFilter({
}) {
  return (
    <Form.Group className="mb-0" >
    </Form.Group>
  );
}
function fuzzyTextFilterFn(rows, id, filterValue) {
  return matchSorter(rows, filterValue, { keys: [(row) => row.values[id]] });
}

fuzzyTextFilterFn.autoRemove = (val) => !val;

function PlanTable({ columns, data, isLoading}) {

  const defaultColumn = React.useMemo(
    () => ({
      Filter: NoneColumnFilter,
    }),
    []
  );
  const props = useTable(
    {
      columns: columns ? columns : [],
      data: data ? data : [],
      defaultColumn,
    },
    useFilters,
    useGlobalFilter,
    useSortBy,
    usePagination,
    useRowSelect,
  );

  const {
    getTableProps,
    getTableBodyProps,
    headerGroups,
    rows,
    prepareRow,
    setGlobalFilter,
    state,
    page,
    canPreviousPage,
    canNextPage,
    pageOptions,
    pageCount,
    gotoPage,
    nextPage,
    previousPage,
    setPageSize,
    state: { pageIndex, pageSize, globalFilter },
  } = props;

  let toggleRow = '';
  let rowColor = true;

  return (
    <>
      <div className="d-flex align-items-center justify-content-end pb-3">
        <div className="mr-3">
          <Label className="mb-0">Search: </Label>
        </div>
        <div className="mr-3">
          <Input
            type="search"
            value={globalFilter || ""}
            onChange={(e) => setGlobalFilter(e.target.value)}
          />
        </div>
      </div>
      <style>{`
      td,tr,th{
        text-align: center;
      }
      `}</style>
      {isLoading ? (
          <div
            className="d-flex flex-column justify-content-center align-items-center"
            style={{ height: '100px', marginBottom: '20px' }}
          >
            <Spinner
              variant="primary"
              animation="border"
              role="status"
              className="position-absolute"
              style={{ height: "60px", width: "60px" }}
            ></Spinner>
          </div>
        ) :(
      <Table hover {...getTableProps()} >
        <thead>
          {headerGroups.map((headerGroup) => (
            <tr {...headerGroup.getHeaderGroupProps()} >        
              {headerGroup.headers.map((column) => (
                <th {...column.getHeaderProps()} className="plan-details-header">
                  <div
                    className="heading"
                    style={{
                      fontSize: "16px",
                      textTransform: "uppercase",
                      color: "black",
                      fontWeight: "bold",
                    }}
                  >
                    {column.render("Header")}
                  </div>
                  {/* Render the columns filter UI */}
                  <div>{column.canFilter ? column.render("Filter") : null}</div>
                </th>
              ))}
            </tr>
          ))}
        </thead>
        <tbody {...getTableBodyProps()}>
          {page.map((row, i) => {
            prepareRow(row);
            let subject = '';
            if(row.values.topic){
              subject = row.original.moduleName;
            } 
            if (subject !== toggleRow) {
              rowColor = !rowColor;
            }
            toggleRow = subject;
  
            return (
              <tr style={{ backgroundColor: rowColor ? "white":"rgb(221, 221, 221)"}} {...row.getRowProps()} className="plan-details-data">
                {row.cells.map((cell) => {
                  return (
                    <td {...cell.getCellProps()}>{cell.render("Cell")}</td>
                  );
                })}
              </tr>
            );
          })}
        </tbody>
      </Table>)}
      <div className=" d-flex align-items-center justify-content-center">
        <Button
          className="btn-fill btn-sm mx-1"
          onClick={() => gotoPage(0)}
          disabled={!canPreviousPage}
        >
          {"<<"}
        </Button>
        <Button
          className="btn-fill btn-sm mx-1"
          onClick={() => previousPage()}
          disabled={!canPreviousPage}
        >
          {"<"}
        </Button>
        <Button
          className="btn-fill btn-sm mx-1"
          onClick={() => nextPage()}
          disabled={!canNextPage}
        >
          {">"}
        </Button>
        <Button
          className="btn-fill btn-sm mx-1"
          onClick={() => gotoPage(pageCount - 1)}
          disabled={!canNextPage}
        >
          {">>"}
        </Button>
        <span className=" mx-3">
          Page
          <strong>
            {pageIndex + 1} of {pageOptions.length}
          </strong>
        </span>
        <span className="mx-3"> Go to page: </span>
        <Input
          type="number"
          className="mr-3"
          defaultValue={pageIndex + 1}
          onChange={(e) => {
            const page = e.target.value ? Number(e.target.value) - 1 : 0;
            gotoPage(page);
          }}
          style={{ width: "100px" }}
        />
        <div className="">
          <Input
            type="select"
            value={pageSize}
            onChange={(e) => {
              setPageSize(Number(e.target.value));
            }}
          >
            {[10, 20, 30, 40, 50].map((pageSize) => (
              <option key={pageSize} value={pageSize}>
                Show {pageSize}
              </option>
            ))}
          </Input>
        </div>
      </div>
      <br/>
    </>
  );
}

export function filterGreaterThan(rows, id, filterValue) {
  return rows.filter((row) => {
    const rowValue = row.values[id];
    return rowValue >= filterValue;
  });
}

filterGreaterThan.autoRemove = (val) => typeof val !== "number";

export default PlanTable;
