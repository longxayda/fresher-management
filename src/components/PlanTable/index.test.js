import React from "react";
import Enzyme from "enzyme";
import Adapter from "@wojtekmaj/enzyme-adapter-react-17";
import renderer from 'react-test-renderer';
import {shallow , mount, render } from "enzyme";

import PlanTable from "./index";

Enzyme.configure({ adapter: new Adapter()});

describe("Test PlanTable input props", () => {
    test("plan table with empty columns & data", () => {
        const columns = [];
        const data = [];
        const tree = shallow(<PlanTable columns={columns} data={data}/>);
        expect(tree.find('.plan-details-header').exists()).toEqual(false);
    })

    test("plan table with NULL columns & data", () => {
        const columns = null;
        const data = null;
        const tree = shallow(<PlanTable columns={columns} data={data}/>);
        expect(tree.find('.plan-details-header').exists()).toEqual(false);
    })

    test('plan table with valid column & empty data',()=>{
        const columns = [{Header: 'Test'}];
        const data = [];
        const tree = shallow(<PlanTable columns={columns} data={data}/>);
        expect(tree.find('.plan-details-header').exists()).toEqual(true);
        expect(tree.find('.plan-details-data').exists()).toEqual(false);
    })

    test('plan table with empty column & valid data',()=>{
        const columns = [];
        const data = [{test:1}];
        const tree = shallow(<PlanTable columns={columns} data={data}/>);
        expect(tree.find('.plan-details-header').exists()).toEqual(false);
        expect(tree.find('.plan-details-data').exists()).toEqual(false);
    })

    test('plan table with valid column & valid data',()=>{
        const columns = [{Header: 'Test', accessor:'test'}];
        const data = [{test: 1}];
        const tree = shallow(<PlanTable columns={columns} data={data}/>);
        expect(tree.find('.plan-details-header').exists()).toEqual(true);
        expect(tree.find('.plan-details-data').exists()).toEqual(true);
    })
})

describe("Test PlanTable render" , () => {
    test('render plan table with cell function',()=>{
        const columns = [{Header: 'Test-cell', Cell: ({row})=><button className="test-cell-function">test cell</button>,}];
        const data = [{}];
        const tree = renderer.create(<PlanTable columns={columns} data={data}/>).root;
        expect(tree.findByProps({className: "test-cell-function"}).children).toEqual(['test cell']);
    })

    test('render plan table with cell function 2',()=>{
        const columns = [{Header: 'Test-cell', Cell: ({row})=><button className="test-cell-function">{row.original.test}</button>,}];
        const data = [{test: 'test cell 2'}];
        const tree = renderer.create(<PlanTable columns={columns} data={data}/>).root;
        expect(tree.findByProps({className: "test-cell-function"}).children).toEqual(['test cell 2']);
    })
});

test("Test PlanTable snapshot", () => {
    const columns = [{Header: 'Test',accessor: 'test'}];
    const data = [{test: 'test 1'},{test:'test 2'}]
    const component = renderer.create(
        <PlanTable columns={columns} data={data}/>,
      );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
})
