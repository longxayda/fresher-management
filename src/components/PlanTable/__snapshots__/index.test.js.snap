// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test PlanTable snapshot 1`] = `
Array [
  <div
    className="d-flex align-items-center justify-content-end pb-3"
  >
    <div
      className="mr-3"
    >
      <label
        className="mb-0"
      >
        Search: 
      </label>
    </div>
    <div
      className="mr-3"
    >
      <input
        className="form-control"
        onChange={[Function]}
        type="search"
        value=""
      />
    </div>
  </div>,
  <style>
    
      td,tr,th{
        text-align: center;
      }
      
  </style>,
  <table
    className="table table-hover"
    role="table"
  >
    <thead>
      <tr
        role="row"
      >
        <th
          className="plan-details-header"
          colSpan={1}
          role="columnheader"
        >
          <div
            className="heading"
            style={
              Object {
                "display": "flex",
                "fontSize": "16px",
                "fontWeight": "600",
                "justifyContent": "center",
              }
            }
          >
            Test
          </div>
          <div>
            <div
              className="mb-0"
            />
          </div>
        </th>
      </tr>
    </thead>
    <tbody
      role="rowgroup"
    >
      <tr
        className="plan-details-data"
        role="row"
        style={
          Object {
            "backgroundColor": "white",
          }
        }
      >
        <td
          role="cell"
        >
          test 1
        </td>
      </tr>
      <tr
        className="plan-details-data"
        role="row"
        style={
          Object {
            "backgroundColor": "white",
          }
        }
      >
        <td
          role="cell"
        >
          test 2
        </td>
      </tr>
    </tbody>
  </table>,
  <div
    className=" d-flex align-items-center justify-content-center"
  >
    <button
      className="btn-fill mx-1 btn btn-primary"
      disabled={true}
      onClick={[Function]}
      type="button"
    >
      &lt;&lt;
    </button>
    <button
      className="btn-fill mx-1 btn btn-primary"
      disabled={true}
      onClick={[Function]}
      type="button"
    >
      &lt;
    </button>
    <button
      className="btn-fill mx-1 btn btn-primary"
      disabled={true}
      onClick={[Function]}
      type="button"
    >
      &gt;
    </button>
    <button
      className="btn-fill mx-1 btn btn-primary"
      disabled={true}
      onClick={[Function]}
      type="button"
    >
      &gt;&gt;
    </button>
    <span
      className=" mx-3"
    >
      Page
      <strong>
        1
         of 
        1
      </strong>
    </span>
    <span
      className="mx-3"
    >
       Go to page: 
    </span>
    <input
      className="mr-3 form-control"
      defaultValue={1}
      onChange={[Function]}
      style={
        Object {
          "width": "100px",
        }
      }
      type="number"
    />
    <div
      className=""
    >
      <select
        className="form-control"
        onChange={[Function]}
        value={10}
      >
        <option
          value={10}
        >
          Show 
          10
        </option>
        <option
          value={20}
        >
          Show 
          20
        </option>
        <option
          value={30}
        >
          Show 
          30
        </option>
        <option
          value={40}
        >
          Show 
          40
        </option>
        <option
          value={50}
        >
          Show 
          50
        </option>
      </select>
    </div>
  </div>,
  <br />,
]
`;
